phase1-coordinator/src/coordinator_state.rs

```diff
@@ -1892,43 +1893,53 @@ impl CoordinatorState {
+            for contributor_info in self.current_contributors.values_mut() {
+                // If the pending task is in the same chunk with the dropped task
+                // then it should be recomputed
+                let (disposing_tasks, pending_tasks) = contributor_info
+                    .pending_tasks
+                    .iter()
+                    .cloned()
+                    .partition(|task| tasks_by_chunk.get(&task.chunk_id).is_some());
+
+                contributor_info.disposing_tasks = disposing_tasks;
+                contributor_info.pending_tasks = pending_tasks;
+
+                // If completed task is based on the dropped task, it should also be dropped
+                let (disposed_tasks, completed_tasks) =
+                    contributor_info.completed_tasks.iter().cloned().partition(|task| {
+                        if let Some(contribution_id) = tasks_by_chunk.get(&task.chunk_id) {
+                            *contribution_id < task.contribution_id
+                        } else {
+                            false
+                        }
+                    });
+                contributor_info.completed_tasks = completed_tasks;
+                contributor_info.disposed_tasks.extend(disposed_tasks);
 
-                    // Move the completed tasks to the disposed tasks.
-                    contributor_info.disposed_tasks = contributor_info.assigned_tasks.clone();
-                    contributor_info
-                        .disposed_tasks
-                        .extend(contributor_info.completed_tasks.iter());
-                    contributor_info.assigned_tasks = LinkedList::new();
-                    contributor_info.completed_tasks = LinkedList::new();
+                all_disposed_tasks.extend(contributor_info.disposed_tasks.iter());
 
-                    // Add the disposed tasks to the cumulative disposed tasks.
-                    disposed_tasks.extend(contributor_info.disposed_tasks.iter());
+                let mut tasks_to_exclude: HashSet<u64> =
+                    HashSet::from_iter(contributor_info.completed_tasks.iter().map(|task| task.chunk_id));
+                tasks_to_exclude.extend(contributor_info.pending_tasks.iter().map(|task| task.chunk_id));
 
-                    // Reassign tasks for the affected contributor.
-                    contributor_info.assigned_tasks =
-                        initialize_tasks(contributor_info.bucket_id, number_of_chunks, number_of_contributors);
+                let new_assigned_tasks =
+                    initialize_tasks(contributor_info.bucket_id, number_of_chunks, number_of_contributors);
 
-                    break;
-                }
+                // Reassign tasks for the affected contributor.
+                contributor_info.assigned_tasks = new_assigned_tasks
+                    .into_iter()
+                    .filter(|task| !tasks_to_exclude.contains(&task.chunk_id))
+                    .collect();
```

Comment:
Can you combine these operations for easier readability?

Reply:
You mean to remove intermediate variable `new_assigned_tasks`?

Reply:
Indeed!